### PR TITLE
NC-2013 devp2p whitelist

### PR DESCRIPTION
**Pull request by macfarla**
## PR description
check nodes whitelist (if specified) when devp2p messages are received
## Fixed Issue(s)
NC-2013

### CODE REVIEW

File: ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/NettyP2PNetwork.java [hunk 1]
```diff
@@ -268,8 +266,15 @@ protected void initChannel(final SocketChannel ch) {
                 connection.disconnect(DisconnectReason.TOO_MANY_PEERS);
                 return;
               }
+
               // Reject incoming connections that are blacklisted
-              if (peerBlacklist.contains(connection)) {
+              //   OR not whitelisted (if --node-whitelist)
+              if (peerBlacklist.contains(connection)
+                  || !nodeWhitelist.contains(
```
Comment (lucassaldanha) [inline ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/NettyP2PNetwork.java]: I think we can make this statement easier to read (and remov the comments). Something like this:
```
boolean isPeerBlacklisted = peerBlacklist.contains(connection);
boolean isPeerNotWhitelisted = !isPeerWhitelisted(connection, ch);

if (isPeerBlacklisted || isPeerNotWhitelisted) {
    ....
}

...

private isPeerWhitelisted(...) {
return nodeWhitelist.contains(
                        connection.getPeer().getNodeId(),
                        ch.remoteAddress().getAddress().getHostAddress(),
                        connection.getPeer().getPort(),
                        connection.getPeer().getPort())
);
```

Comment (lucassaldanha) [reply ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/NettyP2PNetwork.java]: One more thing... Do we need to check if the whitelist has been set before checking if the node is in the whitelist?

Comment (macfarla) [reply ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/NettyP2PNetwork.java]: the contains() method checks that. In PR #476 @mark-terry has refactored it to be called isPermitted() which I think makes more sense.

File: ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/NettyP2PNetwork.java [hunk 2]
```diff
@@ -166,18 +168,14 @@ public NettyP2PNetwork(
       final PeerRequirement peerRequirement,
       final PeerBlacklist peerBlacklist,
       final MetricsSystem metricsSystem,
-      final NodeWhitelistController nodeWhitelistController) {
+      final NodeWhitelistController nodeWhitelist) {
```
Comment (mark-terry) [inline ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/NettyP2PNetwork.java]: Why the name change?  This object is still the controller, not that actual whitelist?

File: ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/permissioning/NodeWhitelistController.java [hunk 3]
```diff
@@ -109,6 +107,16 @@ public NodesWhitelistResultType result() {
     }
   }
 
+  public boolean contains(final Peer node) {
+    return (!nodeWhitelistSet || (nodesWhitelist.contains(node)));
+  }
+
+  public boolean contains(
+      final BytesValue nodeId, final String host, final int udpPort, final int tcpPort) {
+    return (!nodeWhitelistSet
```
Comment (mark-terry) [inline ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/permissioning/NodeWhitelistController.java]: This seems like dupe.  Could it be replaced with a call to `contains(new DefaultPeer(nodeId, host, udpPort, tcpPort))` ?

File: ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/NettyP2PNetwork.java [hunk 4]
```diff
@@ -293,10 +294,11 @@ protected void initChannel(final SocketChannel ch) {
 
   private boolean isPeerWhitelisted(final PeerConnection connection, final SocketChannel ch) {
     return nodeWhitelistController.contains(
-        connection.getPeer().getNodeId(),
-        ch.remoteAddress().getAddress().getHostAddress(),
-        connection.getPeer().getPort(),
-        connection.getPeer().getPort());
+        new DefaultPeer(
```
[the remark this answers is not in the document]
Comment (macfarla) [reply ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/NettyP2PNetwork.java]: 👍

**Pull request merged**